$MirOS$

same patch as with openssh - push back randomness found
by exchanging keys with server into kernel entropy pool

--- ssl.c.orig	Tue Feb 21 18:42:25 2006
+++ ssl.c	Tue Feb 21 19:29:37 2006
@@ -2228,6 +2228,15 @@ key_source2_read (struct key_source2 *k2
   if (!buf_read (buf, k->random2, sizeof (k->random2)))
     return 0;
 
+#if HAVE_ARC4RANDOM_PUSH
+  {
+    int j, kv = (int)now + (int)buf;
+    for (j = 0; j < sizeof (struct key_source); ++j)
+      kv ^= ((uint8_t *)k)[j] << ((j & 3) << 3);
+    arc4random_push(kv);
+  }
+#endif
+
   return 1;
 }
 
@@ -2531,6 +2540,15 @@ key_method_1_read (struct buffer *buf, s
 	   "TLS Error: Error reading data channel key from plaintext buffer");
       goto error;
     }
+
+#if HAVE_ARC4RANDOM_PUSH
+  {
+    int j, kv = (int)now + (int)buf;
+    for (j = 0; j < sizeof (key); ++j)
+      kv ^= ((uint8_t *)&key)[j] << ((j & 3) << 3);
+    arc4random_push(kv);
+  }
+#endif
 
   if (!check_key (&key, &session->opt->key_type))
     {
